File: Problems/0100_0199/0117_Populating_Next_Right_Pointers_in_Each_Node_II/Project_Go/solution/OperateNode.go
package solution

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

func createNode(flds string) *Node {
	return createSubNode(strings.Split(flds, ","), 0, 0)
}

func createSubNode(flds []string, depth int, pos int) *Node {
	if len(flds) == 0 {
		return nil
	}

	curPos := 0
	for i := 0; i < depth; i++ {
		curPos += int(math.Pow(2.0, float64(i)))
	}

	if curPos+pos > len(flds)-1 {
		return nil
	}

	if flds[curPos+pos] == "null" {
		return nil
	}

	val, err := strconv.Atoi(flds[curPos+pos])

	if err != nil {
		fmt.Printf("setNode() Error ... flds[%d] = %s", curPos+pos, flds[curPos+pos])
		return nil
	}

	node := new(Node)
	node.Val = val
	node.Left = createSubNode(flds, depth+1, 2*pos)
	node.Right = createSubNode(flds, depth+1, 2*pos+1)
	return node
}

var resultList = []string{}

func TreeToStaircaseString(node *Node) string {
	resultList = []string{}

	resultStr := treeToStaircaseSubString(node, 0)
	resultList = nil

	return resultStr
}

func treeToStaircaseSubString(node *Node, n int) string {
	if node == nil {
		return ""
	}

	if len(resultList) <= n {
		resultList = append(resultList, "("+strconv.Itoa(node.Val)+")")
	} else {
		resultList[n] = resultList[n] + ",(" + strconv.Itoa(node.Val) + ")"
	}

	if node.Left != nil {
		treeToStaircaseSubString(node.Left, n+1)
	}
	if node.Right != nil {
		treeToStaircaseSubString(node.Right, n+1)
	}

	resultStr := ""
	for _, s := range resultList {
		resultStr += s + "\n"
	}

	return resultStr
}

func TreeToStaircaseString_with_next(node *Node) string {
	resultList = []string{}

	resultStr := treeToStaircaseSubString_with_next(node, 0)
	resultList = nil

	return resultStr
}

func treeToStaircaseSubString_with_next(node *Node, n int) string {
	if node == nil {
		return ""
	}

	if len(resultList) <= n {
		resultList = append(resultList, "("+strconv.Itoa(node.Val)+")")
	} else {
		resultList[n] = resultList[n] + ",(" + strconv.Itoa(node.Val) + ")"
	}

	if node.Next == nil {
		resultList[n] = resultList[n] + ",(#)"
	}

	if node.Left != nil {
		treeToStaircaseSubString_with_next(node.Left, n+1)
	}
	if node.Right != nil {
		treeToStaircaseSubString_with_next(node.Right, n+1)
	}

	resultStr := ""
	for _, s := range resultList {
		resultStr += s + "\n"
	}

	return resultStr
}

func Tree2str(node *Node) string {
	if node == nil {
		return ""
	}

	resultStr := strconv.Itoa(node.Val)

	if node.Left == nil && node.Right == nil {
		return resultStr
	}

	resultStr += "(" + Tree2str(node.Left) + ")"
	if node.Right != nil {
		resultStr += "(" + Tree2str(node.Right) + ")"
	}

	return resultStr
}

File: Problems/0100_0199/0117_Populating_Next_Right_Pointers_in_Each_Node_II/Project_Python3/Node/OperateNode.py
from .Node import Node

class OperateNode:
    def createNode(self, flds):
        return self.createSubNode(flds.split(","), 0, 0)

    def createSubNode(self, flds, depth, pos):
        if len(flds) <= 0:
            return None

        cur_pos = 0
        for i in range(depth):
            cur_pos += 2 ** i
        
        if cur_pos + pos > len(flds) - 1:
            return None

        if flds[cur_pos + pos] == 'null':
            return None

        node = Node(int(flds[cur_pos + pos]))
        node.left = self.createSubNode(flds, depth + 1, 2*pos)
        node.right = self.createSubNode(flds, depth + 1, 2*pos + 1)

        return node

    def treeToStaircaseString(self, node):
        self.resultList = []
        self.getStaircaseSubString(node, 0)
        return self.resultListToString()

    def getStaircaseSubString(self, node, n):
        if node is None:
            return
        if len(self.resultList) <= n:
            self.resultList.append("(" + str(node.val) + ")")
        else:
            self.resultList[n] += ",(" + str(node.val) + ")"
        if node.left is not None:
            self.getStaircaseSubString(node.left, n + 1)
        if node.right is not None:
            self.getStaircaseSubString(node.right, n + 1)
        return

    def treeToStaircaseString_with_next(self, node):
        self.resultList = []
        self.getStaircaseSubString_with_next(node, 0)
        return self.resultListToString()

    def getStaircaseSubString_with_next(self, node, n):
        if node is None:
            return
        if len(self.resultList) <= n:
            self.resultList.append("(" + str(node.val) + ")")
        else:
            self.resultList[n] += ",(" + str(node.val) + ")"
        if node.next is None:
            self.resultList[n] += ",(#)"
        if node.left is not None:
            self.getStaircaseSubString_with_next(node.left, n + 1)
        if node.right is not None:
            self.getStaircaseSubString_with_next(node.right, n + 1)
        return

    def resultListToString(self):
        resultStr = ""
        for i in range(len(self.resultList)):
            resultStr += self.resultList[i] + "\n"
        self.resultList.clear()
        return resultStr

    def tree2str(self, t):
        """
        :type t: TreeNode
        :rtype: str
        """
        if t is None:
            return ""

        resultStr = str(t.val)

        if t.left is None and t.right is None:
            return resultStr

        resultStr += "(" + self.tree2str(t.left) + ")"
        if t.right is not None:
            resultStr += "(" + self.tree2str(t.right) + ")"

        return resultStr
